### ISSUE TITLE
ZIO 2.0: ZIO.forever exhausts heap memory

### ISSUE THREAD

**Issue by ajaychandran** [zio/zio#5870]
[Scastie](https://scastie.scala-lang.org/XWkgDH2GS8mZmAP2IHwicg)
```scala
import zio._

Runtime.default.unsafeRun(ZIO.unit.forever)
```
```sbt
[warn] In the last 10 seconds, 9.307 (96.8%) were spent in GC. [Heap: 0.17GB free of 0.89GB, max 0.89GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.
```

Redefining the combinator as shown below works.
```scala
final def forever(implicit trace: ZTraceElement): ZIO[R, E, A] = {

  lazy val loop: ZIO[R, E, A] = new ZIO.FlatMap(self <* ZIO.yieldNow, (_: A) => loop, trace)

  loop
}
```

Also, combinators like `repeatWhile` can exhibit this problem.

**Comment by adamgfraser** [comment]
@ajaychandran Thanks for reporting! Will take a look.

**Comment by adamgfraser** [comment]
Closed by #5985.
